File: test/sandbox.test.ts
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

// Mutable platform override so the seatbelt guard-rails are testable on any OS.
// Fallback is process.platform, NOT a node:os import (that would resolve to this
// very mock and recurse).
let platformOverride: string | null = null;
vi.mock('node:os', async (importOriginal) => {
  const actual = await importOriginal<typeof import('node:os')>();
  return {
    ...actual,
    platform: () => platformOverride ?? actual.platform(),
  };
});

// Capture what spawnCli ultimately hands to execFile.
type CapturedCall = { cmd: string; args: string[]; env?: Record<string, string> };
const captured: CapturedCall[] = [];
const execFileMock = vi.fn(
  (
    cmd: string,
    args: string[],
    opts: { env?: Record<string, string> },
    cb: (error: null, stdout: string, stderr: string) => void,
  ) => {
    captured.push({ cmd, args, env: opts?.env });
    setImmediate(() => cb(null, JSON.stringify({ result: 'ok', session_id: 's1' }), ''));
    return undefined as never;
  },
);
vi.mock('node:child_process', () => ({
  execFile: (...args: unknown[]) => (execFileMock as unknown)(...(args as [])),
}));

// See test/workers.test.ts: ambient DEVAGENT_NO_PROGRESS_TIMEOUT_MS from the
// operator's shell routes spawns through spawnCliStreaming -> spawn, which
// this execFile-only mock does not define. Pin the env for determinism.
beforeEach(() => {
  delete process.env.DEVAGENT_NO_PROGRESS_TIMEOUT_MS;
});

// Controllable DNS for allowlist-resolution cases.
const dnsLookupMock = vi.fn();
vi.mock('node:dns', async (importOriginal) => {
  const actual = await importOriginal<typeof import('node:dns')>();
  return {
    ...actual,
    promises: {
      ...actual.promises,
      lookup: (...args: unknown[]) =>
        (dnsLookupMock as unknown)(...(args as [])),
    },
  };
});

import { readFileSync } from 'node:fs';
import { ClaudeCodeAdapter } from '../src/workers/claude-code.js';
import { OpenCodeAdapter } from '../src/workers/opencode.js';
import {
  buildSeatbeltProfile,
  prepareWorkerSpawn,
  resolveNetworkAllowlist,
  sanitizeWorkerEnv,
} from '../src/workers/sandbox.js';
import type { SpawnCliOptions } from '../src/workers/spawn-utils.js';

const SPAWN_OPTS: SpawnCliOptions = { cwd: '/repo/worktree', timeoutMs: 5_000 };

describe('sanitizeWorkerEnv', () => {
  const baseEnv = {
    PATH: '/usr/bin',
    HOME: '/Users/t',
    ANTHROPIC_API_KEY: 'sk-ant-x',
    GITHUB_TOKEN: 'ghp-secret',
    NPM_TOKEN: 'npm-secret',
    AWS_SECRET_ACCESS_KEY: 'aws-secret',
    MY_SERVICE_PASSWORD: 'pw',
    SLACK_CLIENT_SECRET: 'slack',
    STRIPE_API_KEY: 'sk-live',
  };

  it('strips secret-shaped vars regardless of provider prefix', () => {
    const { env, stripped } = sanitizeWorkerEnv(baseEnv);
    // GITHUB_TOKEN is allowlisted (worker-side gh/curl api.github.com calls —
    // research crawls exhausted the anonymous 60 req/h IP budget, 2026-09-02).
    expect(stripped.sort()).toEqual([
      'AWS_SECRET_ACCESS_KEY',
      'MY_SERVICE_PASSWORD',
      'NPM_TOKEN',
      'SLACK_CLIENT_SECRET',
      'STRIPE_API_KEY',
    ]);
    expect(env.GITHUB_TOKEN).toBe('ghp-secret');
    expect(env.NPM_TOKEN).toBeUndefined();
    expect(env.AWS_SECRET_ACCESS_KEY).toBeUndefined();
  });

  it('keeps process basics and LLM auth vars the worker CLIs need', () => {
    const { env, stripped } = sanitizeWorkerEnv(baseEnv);
    expect(env.PATH).toBe('/usr/bin');
    expect(env.HOME).toBe('/Users/t');
    expect(env.ANTHROPIC_API_KEY).toBe('sk-ant-x');
    expect(stripped).not.toContain('ANTHROPIC_API_KEY');
  });

  it('does not mutate the caller environment', () => {
    sanitizeWorkerEnv(baseEnv);
    expect(baseEnv.GITHUB_TOKEN).toBe('ghp-secret');
  });

  describe('DEVAGENT_WORKER_ENV_ALLOWLIST override', () => {
    beforeEach(() => {
      vi.stubEnv('DEVAGENT_WORKER_ENV_ALLOWLIST', 'STRIPE_API_KEY,CUSTOM_TOKEN');
    });
    afterEach(() => {
      vi.unstubAllEnvs();
    });

    it('keeps explicitly allowlisted names even when secret-shaped', () => {
      const { env, stripped } = sanitizeWorkerEnv({
        ...baseEnv,
        CUSTOM_TOKEN: 'needed-by-worker',
      });
      expect(env.STRIPE_API_KEY).toBe('sk-live');
      expect(env.CUSTOM_TOKEN).toBe('needed-by-worker');
      expect(stripped).not.toContain('STRIPE_API_KEY');
      expect(stripped).not.toContain('CUSTOM_TOKEN');
      // other secrets still stripped (GITHUB_TOKEN is baseline-allowlisted)
      expect(env.NPM_TOKEN).toBeUndefined();
    });
  });
});

describe('buildSeatbeltProfile', () => {
  it('denies writes by default and allowlists cwd, temp dirs, agent config home', () => {
    const profile = buildSeatbeltProfile({ writablePaths: ['/repo/wt'], network: 'allow' });
    expect(profile).toContain('(deny file-write*)');
    expect(profile).toContain('(subpath "/repo/wt")');
    expect(profile).toContain('(subpath "/private/tmp")');
    expect(profile).toContain('(subpath "/private/var/folders")');
    expect(profile).toMatch(/\(subpath "[^"]*\/\.claude"\)/);
  });

  it('expresses network policy as an explicit clause', () => {
    const profile = buildSeatbeltProfile({ writablePaths: [], network: 'allow' });
    expect(profile).toContain('(allow network)');
  });

  it('deny policy emits (deny network*) after the default allow, last-match-wins', () => {
    const profile = buildSeatbeltProfile({ writablePaths: [], network: 'deny' });
    expect(profile).toContain('(deny network*)');
    expect(profile.indexOf('(allow default)')).toBeLessThan(profile.indexOf('(deny network*)'));
    expect(profile).not.toContain('(allow network)');
  });

  it('network deny composes with the write allowlist instead of replacing it', () => {
    const profile = buildSeatbeltProfile({ writablePaths: ['/repo/wt'], network: 'deny' });
    expect(profile).toContain('(deny file-write*)');
    expect(profile).toContain('(subpath "/repo/wt")');
    expect(profile).toContain('(deny network*)');
  });

  it('allowlist mode denies all sockets, then re-allows each endpoint after it', () => {
    const endpoints = ['140.82.121.3:443', '104.16.26.34:443'];
    const profile = buildSeatbeltProfile({
      writablePaths: [],
      network: 'allowlist',
      networkAllowlist: endpoints,
    });
    expect(profile).toContain('(deny network*)');
    // SBPL is last-match-wins: re-allow clauses must come after the blanket deny.
    const denyIdx = profile.indexOf('(deny network*)');
    const allowIdxs = endpoints.map((e) =>
      profile.indexOf(`(allow network-outbound (remote ip "${e}"))`),
    );
    expect(allowIdxs[0]).toBeGreaterThan(-1);
    expect(allowIdxs[1]).toBeGreaterThan(-1);
    expect(denyIdx).toBeLessThan(allowIdxs[0]!);
    expect(allowIdxs[0]!).toBeLessThan(allowIdxs[1]!);
    // one clause per endpoint, no bare network allow
    expect(profile.match(/allow network-outbound/g)).toHaveLength(endpoints.length);
    expect(profile).not.toContain('(allow network)');
  });
});

describe('resolveNetworkAllowlist', () => {
  afterEach(() => {
    dnsLookupMock.mockReset();
  });

  it('resolves hostnames via dns lookup(all) with default port 443', async () => {
    dnsLookupMock.mockResolvedValue([
      { address: '140.82.121.3', family: 4 },
      { address: '2606:50c0:8000::153', family: 6 },
    ]);
    await expect(resolveNetworkAllowlist('api.github.com')).resolves.toEqual([
      '140.82.121.3:443',
      '2606:50c0:8000::153:443',
    ]);
    expect(dnsLookupMock).toHaveBeenCalledWith('api.github.com', { all: true });
  });

  it('applies default port 443 when an entry omits the port', async () => {
    dnsLookupMock.mockResolvedValue([{ address: '108.128.1.5', family: 4 }]);
    await expect(resolveNetworkAllowlist('api.anthropic.com')).resolves.toEqual([
      '108.128.1.5:443',
    ]);
  });

  it('honors explicit ports on hostname entries', async () => {
    dnsLookupMock.mockResolvedValue([{ address: '192.168.1.10', family: 4 }]);
    await expect(resolveNetworkAllowlist('registry.npmjs.org:8443')).resolves.toEqual([
      '192.168.1.10:8443',
    ]);
  });

  it('passes literal IPv4 and IPv6 entries through without touching dns', async () => {
    await expect(resolveNetworkAllowlist('10.0.0.5:8443, fd00::7')).resolves.toEqual([
      '10.0.0.5:8443',
      'fd00::7:443',
    ]);
    expect(dnsLookupMock).not.toHaveBeenCalled();
  });

  it('throws loudly naming an unresolvable host', async () => {
    dnsLookupMock.mockRejectedValue(new Error('lookup failed: ENOTFOUND'));
    await expect(resolveNetworkAllowlist('nope.invalid')).rejects.toThrow(/nope\.invalid/);
    expect(dnsLookupMock).toHaveBeenCalledTimes(1);
  });

  it('throws loudly when every entry is blank', async () => {
    await expect(resolveNetworkAllowlist(' , ,')).rejects.toThrow(
      /DEVAGENT_SANDBOX_NETWORK_ALLOWLIST/,
    );
  });
});

describe('prepareWorkerSpawn', () => {
  beforeEach(() => {
    captured.length = 0;
    vi.unstubAllEnvs();
  });
  afterEach(() => {
    vi.unstubAllEnvs();
    platformOverride = null;
    dnsLookupMock.mockReset();
  });

  it('scrubs the worker env by default and flags replaceEnv', async () => {
    vi.stubEnv('NPM_TOKEN', 'x');
    const prepared = await prepareWorkerSpawn('claude', ['-p', 'hi'], SPAWN_OPTS);
    expect(prepared.cmd).toBe('claude');
    expect(prepared.opts.replaceEnv).toBe(true);
    expect(prepared.opts.env!.NPM_TOKEN).toBeUndefined();
    expect(prepared.strippedEnv).toContain('NPM_TOKEN');
  });

  it('lets caller-provided opts.env win over the scrubbed base', async () => {
    vi.stubEnv('OPENAI_API_KEY', 'parent-key');
    const prepared = await prepareWorkerSpawn('claude', ['-p'], {
      ...SPAWN_OPTS,
      env: { OPENAI_API_KEY: 'override-key' },
    });
    expect(prepared.opts.env!.OPENAI_API_KEY).toBe('override-key');
  });

  it('fails loudly when seatbelt is requested on non-darwin', async () => {
    platformOverride = 'linux';
    vi.stubEnv('DEVAGENT_SANDBOX', 'seatbelt');
    await expect(prepareWorkerSpawn('claude', ['-p'], SPAWN_OPTS)).rejects.toThrow(/darwin/);
  });

  it.runIf(process.platform === 'darwin')(
    'wraps the command with sandbox-exec and writes the profile when enabled',
    async () => {
      vi.stubEnv('DEVAGENT_SANDBOX', 'seatbelt');
      const prepared = await prepareWorkerSpawn('claude', ['-p', 'hi'], SPAWN_OPTS);
      expect(prepared.cmd).toBe('/usr/bin/sandbox-exec');
      expect(prepared.args[0]).toBe('-f');
      const profileText = readFileSync(prepared.args[1], 'utf8');
      expect(profileText).toContain('(deny file-write*)');
      expect(profileText).toContain('(subpath "/repo/worktree")');
      expect(prepared.args.slice(2)).toEqual(['claude', '-p', 'hi']);
      expect(prepared.opts.replaceEnv).toBe(true);
    },
  );

  it.runIf(process.platform === 'darwin')(
    'DEVAGENT_SANDBOX_NETWORK=deny lands in the generated profile; unset keeps allow',
    async () => {
      vi.stubEnv('DEVAGENT_SANDBOX', 'seatbelt');
      const allowPrepared = await prepareWorkerSpawn('claude', ['-p'], SPAWN_OPTS);
      expect(readFileSync(allowPrepared.args[1], 'utf8')).toContain('(allow network)');

      vi.stubEnv('DEVAGENT_SANDBOX_NETWORK', 'deny');
      const denyPrepared = await prepareWorkerSpawn('claude', ['-p'], SPAWN_OPTS);
      const denyProfile = readFileSync(denyPrepared.args[1], 'utf8');
      expect(denyProfile).toContain('(deny network*)');
      expect(denyProfile).not.toContain('(allow network)');
    },
  );

  it.runIf(process.platform === 'darwin')(
    'throws loudly when allowlist mode has an empty or missing allowlist var',
    async () => {
    platformOverride = 'darwin';
    vi.stubEnv('DEVAGENT_SANDBOX', 'seatbelt');
    vi.stubEnv('DEVAGENT_SANDBOX_NETWORK', 'allowlist');
    vi.stubEnv('DEVAGENT_SANDBOX_NETWORK_ALLOWLIST', '');
    await expect(prepareWorkerSpawn('claude', ['-p'], SPAWN_OPTS)).rejects.toThrow(
      /DEVAGENT_SANDBOX_NETWORK_ALLOWLIST/,
    );
    vi.stubEnv('DEVAGENT_SANDBOX_NETWORK_ALLOWLIST', '   ');
    await expect(prepareWorkerSpawn('claude', ['-p'], SPAWN_OPTS)).rejects.toThrow(
      /DEVAGENT_SANDBOX_NETWORK_ALLOWLIST/,
    );
    },
  );

  it.runIf(process.platform === 'darwin')(
    'allowlist mode writes a profile with deny-then-resolved-endpoint clauses',
    async () => {
      dnsLookupMock.mockImplementation(async (host: string) => {
        if (host === 'api.anthropic.com') {
          return [{ address: '108.128.1.5', family: 4 }];
        }
        return [{ address: '151.101.0.1', family: 4 }];
      });
      platformOverride = 'darwin';
      vi.stubEnv('DEVAGENT_SANDBOX', 'seatbelt');
      vi.stubEnv('DEVAGENT_SANDBOX_NETWORK', 'allowlist');
      vi.stubEnv(
        'DEVAGENT_SANDBOX_NETWORK_ALLOWLIST',
        'api.anthropic.com, registry.npmjs.org:443',
      );
      const prepared = await prepareWorkerSpawn('claude', ['-p'], SPAWN_OPTS);
      expect(prepared.cmd).toBe('/usr/bin/sandbox-exec');
      const profile = readFileSync(prepared.args[1], 'utf8');
      const denyIdx = profile.indexOf('(deny network*)');
      const anthropicIdx = profile.indexOf('(allow network-outbound (remote ip "108.128.1.5:443"))');
      const npmIdx = profile.indexOf('(allow network-outbound (remote ip "151.101.0.1:443"))');
      expect(denyIdx).toBeGreaterThan(-1);
      expect(anthropicIdx).toBeGreaterThan(denyIdx);
      expect(npmIdx).toBeGreaterThan(anthropicIdx);
    },
  );

  it.runIf(process.platform === 'darwin')(
    'allowlist mode surfaces unresolvable hosts with the host named',
    async () => {
      dnsLookupMock.mockRejectedValue(new Error('lookup failed: ENOTFOUND'));
      platformOverride = 'darwin';
      vi.stubEnv('DEVAGENT_SANDBOX', 'seatbelt');
      vi.stubEnv('DEVAGENT_SANDBOX_NETWORK', 'allowlist');
      vi.stubEnv('DEVAGENT_SANDBOX_NETWORK_ALLOWLIST', 'dead.internal:443');
      await expect(prepareWorkerSpawn('claude', ['-p'], SPAWN_OPTS)).rejects.toThrow(
        /dead\.internal/,
      );
    },
  );
});

describe('worker adapters route through the sandbox', () => {
  beforeEach(() => {
    captured.length = 0;
    vi.unstubAllEnvs();
    // Hermetic vs selfbuild-loop.sh's exported DEVAGENT_NO_PROGRESS_TIMEOUT_MS:
    // an armed default routes spawnCli through spawnCliStreaming (unmocked
    // spawn here). Watchdog arming is covered in watchdog-health.test.ts.
    delete process.env.DEVAGENT_NO_PROGRESS_TIMEOUT_MS;
  });
  afterEach(() => {
    vi.unstubAllEnvs();
  });

  it('claude-code spawn reaches execFile scrubbed', async () => {
    vi.stubEnv('NPM_TOKEN', 'x');
    const adapter = new ClaudeCodeAdapter();
    const result = await adapter.spawn({
      prompt: 'do thing',
      cwd: '/repo/wt',
      timeoutMs: 5_000,
    });
    expect(result.exitCode).toBe(0);
    expect(captured).toHaveLength(1);
    expect(captured[0].cmd).toBe('claude');
    expect(captured[0].env!.NPM_TOKEN).toBeUndefined();
    expect(captured[0].env!.PATH).toBeDefined();
  });

  it('opencode spawn reaches execFile scrubbed', async () => {
    vi.stubEnv('NPM_TOKEN', 'x');
    const adapter = new OpenCodeAdapter();
    const result = await adapter.spawn({
      prompt: 'do thing',
      cwd: '/repo/wt',
      timeoutMs: 5_000,
    });
    expect(result.exitCode).toBe(0);
    expect(captured).toHaveLength(1);
    expect(captured[0].cmd).toBe('opencode');
    expect(captured[0].args.slice(0, 3)).toEqual(['run', '--format', 'json']);
    expect(captured[0].env!.NPM_TOKEN).toBeUndefined();
  });
});

File: internal/workers/sandbox_test.go
// Go port of test/sandbox.test.ts: env scrub with the
// DEVAGENT_WORKER_ENV_ALLOWLIST override, seatbelt profile generation,
// network allowlist resolution, and prepareWorkerSpawn behavior.

package workers

import (
	"regexp"
	"strings"
	"testing"
)

// Fake provider-shaped values, assembled at runtime so this file never
// contains a usable credential-shaped literal (SanitizeWorkerEnv only
// matches on variable NAMES — the values are opaque).
const (
	fakeAnthropicKey = "sk-" + "ant-x"
	fakeGithubToken  = "ghp-" + "secret"
	fakeStripeKey    = "sk-" + "live"
	fakeNpmToken     = "npm-" + "secret"
	fakeAwsSecret    = "aws-" + "secret"
	fakeXaiKey       = "xai-" + "secret"
)

func sandboxBaseEnv() map[string]string {
	return map[string]string{
		"PATH":                  "/usr/bin",
		"HOME":                  "/Users/t",
		"ANTHROPIC_API_KEY":     fakeAnthropicKey,
		"GITHUB_TOKEN":          fakeGithubToken,
		"NPM_TOKEN":             fakeNpmToken,
		"AWS_SECRET_ACCESS_KEY": fakeAwsSecret,
		"MY_SERVICE_PASSWORD":   "pw",
		"SLACK_CLIENT_SECRET":   "slack",
		"STRIPE_API_KEY":        fakeStripeKey,
		"XAI_API_KEY":           fakeXaiKey,
	}
}

func TestSanitizeWorkerEnv_SecretShapedStripped(t *testing.T) {
	res := SanitizeWorkerEnv(sandboxBaseEnv(), nil)
	want := []string{"AWS_SECRET_ACCESS_KEY", "MY_SERVICE_PASSWORD", "NPM_TOKEN", "SLACK_CLIENT_SECRET", "STRIPE_API_KEY"}
	assertSetEqual(t, res.Stripped, want)
	if res.Env["GITHUB_TOKEN"] != fakeGithubToken {
		// GITHUB_TOKEN is allowlisted (worker-side gh/curl api.github.com
		// calls — research crawls exhausted the anonymous 60 req/h IP
		// budget, 2026-09-02).
		t.Fatalf("GITHUB_TOKEN = %q", res.Env["GITHUB_TOKEN"])
	}
	if _, ok := res.Env["NPM_TOKEN"]; ok {
		t.Fatal("NPM_TOKEN must be stripped")
	}
	if _, ok := res.Env["AWS_SECRET_ACCESS_KEY"]; ok {
		t.Fatal("AWS_SECRET_ACCESS_KEY must be stripped")
	}
}

func TestSanitizeWorkerEnv_KeepsProcessBasicsAndLLMAuth(t *testing.T) {
	res := SanitizeWorkerEnv(sandboxBaseEnv(), nil)
	if res.Env["PATH"] != "/usr/bin" || res.Env["HOME"] != "/Users/t" {
		t.Fatalf("env basics = %v", res.Env)
	}
	if res.Env["ANTHROPIC_API_KEY"] != fakeAnthropicKey {
		t.Fatal("ANTHROPIC_API_KEY must be kept")
	}
	// XAI_API_KEY is allowlisted for the grok worker (FR-GROK-01): the
	// /_API_KEY$/ scrubber would otherwise strip the CLI's only env auth.
	if res.Env["XAI_API_KEY"] != fakeXaiKey {
		t.Fatal("XAI_API_KEY must be kept")
	}
	for _, n := range res.Stripped {
		if n == "ANTHROPIC_API_KEY" || n == "XAI_API_KEY" {
			t.Fatalf("allowlisted var stripped: %q", n)
		}
	}
}

func TestSanitizeWorkerEnv_DoesNotMutateCallerEnv(t *testing.T) {
	base := sandboxBaseEnv()
	SanitizeWorkerEnv(base, nil)
	if base["GITHUB_TOKEN"] != fakeGithubToken {
		t.Fatal("caller env mutated")
	}
}

func TestSanitizeWorkerEnv_AllowlistOverride(t *testing.T) {
	t.Setenv("DEVAGENT_WORKER_ENV_ALLOWLIST", "STRIPE_API_KEY,CUSTOM_TOKEN")
	base := sandboxBaseEnv()
	base["CUSTOM_TOKEN"] = "needed-by-worker"
	res := SanitizeWorkerEnv(base, nil)
	if res.Env["STRIPE_API_KEY"] != fakeStripeKey {
		t.Fatal("explicitly allowlisted secret-shape name must be kept")
	}
	if res.Env["CUSTOM_TOKEN"] != "needed-by-worker" {
		t.Fatal("CUSTOM_TOKEN must be kept")
	}
	for _, n := range res.Stripped {
		if n == "STRIPE_API_KEY" || n == "CUSTOM_TOKEN" {
			t.Fatalf("allowlisted var stripped: %q", n)
		}
	}
	// other secrets still stripped (GITHUB_TOKEN is baseline-allowlisted)
	if _, ok := res.Env["NPM_TOKEN"]; ok {
		t.Fatal("NPM_TOKEN must still be stripped")
	}
}

func TestBuildSeatbeltProfile_WritePolicy(t *testing.T) {
	profile := BuildSeatbeltProfile(SandboxPolicy{WritablePaths: []string{"/repo/wt"}, Network: "allow"})
	if !contains(profile, "(deny file-write*)") {
		t.Fatal("missing blanket write deny")
	}
	if !contains(profile, `(subpath "/repo/wt")`) {
		t.Fatal("missing cwd subpath")
	}
	if !contains(profile, `(subpath "/private/tmp")`) || !contains(profile, `(subpath "/private/var/folders")`) {
		t.Fatal("missing temp-dir subpaths")
	}
	if !regexp.MustCompile(`\(subpath "[^"]*\/\.claude"\)`).MatchString(profile) {
		t.Fatal("missing agent config home subpath")
	}
}

func TestBuildSeatbeltProfile_NetworkPolicy(t *testing.T) {
	allow := BuildSeatbeltProfile(SandboxPolicy{Network: "allow"})
	if !contains(allow, "(allow network)") {
		t.Fatal("allow policy must emit (allow network)")
	}

	deny := BuildSeatbeltProfile(SandboxPolicy{Network: "deny"})
	if !contains(deny, "(deny network*)") {
		t.Fatal("deny policy must emit (deny network*)")
	}
	// SBPL is last-match-wins: the deny clause must come after (allow default).
	if strings.Index(deny, "(allow default)") >= strings.Index(deny, "(deny network*)") {
		t.Fatal("deny must come after the default allow")
	}
	if contains(deny, "(allow network)") {
		t.Fatal("deny policy must not carry a bare network allow")
	}

	// deny composes with the write allowlist instead of replacing it
	composed := BuildSeatbeltProfile(SandboxPolicy{WritablePaths: []string{"/repo/wt"}, Network: "deny"})
	if !contains(composed, "(deny file-write*)") || !contains(composed, `(subpath "/repo/wt")`) || !contains(composed, "(deny network*)") {
		t.Fatalf("composed profile = %s", composed)
	}

	// allowlist mode: deny all sockets, then re-allow each endpoint after it.
	endpoints := []string{"140.82.121.3:443", "104.16.26.34:443"}
	profile := BuildSeatbeltProfile(SandboxPolicy{Network: "allowlist", NetworkAllowlist: endpoints})
	if !contains(profile, "(deny network*)") {
		t.Fatal("allowlist mode must start with the blanket deny")
	}
	denyIdx := strings.Index(profile, "(deny network*)")
	prev := denyIdx
	for _, e := range endpoints {
		clause := `(allow network-outbound (remote ip "` + e + `"))`
		idx := strings.Index(profile, clause)
		if idx < 0 {
			t.Fatalf("missing re-allow clause for %s", e)
		}
		if idx < prev {
			t.Fatalf("re-allow for %s must come after the previous clause", e)
		}
		prev = idx
	}
	if got := strings.Count(profile, "allow network-outbound"); got != len(endpoints) {
		t.Fatalf("re-allow count = %d, want %d", got, len(endpoints))
	}
	if contains(profile, "(allow network)") {
		t.Fatal("allowlist mode must not carry a bare network allow")
	}
}

func TestResolveNetworkAllowlist(t *testing.T) {
	t.Run("literal IPv4 and IPv6 entries pass through without DNS", func(t *testing.T) {
		eps, err := ResolveNetworkAllowlist("10.0.0.5:8443, fd00::7")
		if err != nil {
			t.Fatal(err)
		}
		// fd00::7 is a bare IPv6 literal; JoinHostPort brackets it —
		// mirroring a valid host:port endpoint for SBPL.
		if len(eps) != 2 || eps[0] != "10.0.0.5:8443" {
			t.Fatalf("endpoints = %v", eps)
		}
	})
	t.Run("bracketed IPv6 with explicit port", func(t *testing.T) {
		eps, err := ResolveNetworkAllowlist("[fd00::7]:8443")
		if err != nil {
			t.Fatal(err)
		}
		if eps[0] != "[fd00::7]:8443" && eps[0] != "fd00::7:8443" {
			t.Fatalf("endpoints = %v", eps)
		}
	})
	t.Run("throws loudly when every entry is blank", func(t *testing.T) {
		_, err := ResolveNetworkAllowlist(" , ,")
		if err == nil || !contains(err.Error(), "DEVAGENT_SANDBOX_NETWORK_ALLOWLIST") {
			t.Fatalf("err = %v", err)
		}
	})
	t.Run("throws loudly naming an unresolvable host", func(t *testing.T) {
		// .invalid is guaranteed unresolvable per RFC 2606.
		_, err := ResolveNetworkAllowlist("nope.invalid")
		if err == nil || !contains(err.Error(), "nope.invalid") {
			t.Fatalf("err = %v", err)
		}
	})
	t.Run("resolves a hostname via DNS with default port 443", func(t *testing.T) {
		// Network-dependent resolution on localhost DNS for localhost.
		eps, err := ResolveNetworkAllowlist("localhost")
		if err != nil {
			t.Skipf("dns unavailable: %v", err)
		}
		if len(eps) == 0 {
			t.Fatal("expected at least one endpoint")
		}
		for _, ep := range eps {
			if !hasSuffix(ep, ":443") {
				t.Fatalf("endpoint %q must default to port 443", ep)
			}
		}
	})
}

func TestPrepareWorkerSpawn(t *testing.T) {
	spawnOpts := SpawnCliOptions{Dir: "/repo/worktree", TimeoutMs: 5_000}

	t.Run("scrubs the worker env by default and flags replaceEnv", func(t *testing.T) {
		t.Setenv("NPM_TOKEN", "x")
		prepared, err := PrepareWorkerSpawn("claude", []string{"-p", "hi"}, spawnOpts)
		if err != nil {
			t.Fatal(err)
		}
		if prepared.Cmd != "claude" {
			t.Fatalf("cmd = %q", prepared.Cmd)
		}
		if !prepared.Opts.ReplaceEnv {
			t.Fatal("replaceEnv must be flagged")
		}
		if _, ok := prepared.Opts.Env["NPM_TOKEN"]; ok {
			t.Fatal("NPM_TOKEN must be scrubbed")
		}
		found := false
		for _, n := range prepared.StrippedEnv {
			if n == "NPM_TOKEN" {
				found = true
			}
		}
		if !found {
			t.Fatalf("strippedEnv = %v", prepared.StrippedEnv)
		}
	})

	t.Run("lets caller-provided env win over the scrubbed base", func(t *testing.T) {
		parentKey := "parent" + "-key"
		overrideKey := "override" + "-key"
		t.Setenv("OPENAI_API_KEY", parentKey)
		prepared, err := PrepareWorkerSpawn("claude", []string{"-p"}, SpawnCliOptions{
			Dir: spawnOpts.Dir, TimeoutMs: spawnOpts.TimeoutMs,
			Env: map[string]string{"OPENAI_API_KEY": overrideKey},
		})
		if err != nil {
			t.Fatal(err)
		}
		if prepared.Opts.Env["OPENAI_API_KEY"] != overrideKey {
			t.Fatalf("env = %q", prepared.Opts.Env["OPENAI_API_KEY"])
		}
	})

	t.Run("fails loudly when seatbelt is requested on non-darwin", func(t *testing.T) {
		t.Setenv("DEVAGENT_SANDBOX", "seatbelt")
		orig := platform
		platform = func() string { return "linux" }
		defer func() { platform = orig }()
		_, err := PrepareWorkerSpawn("claude", []string{"-p"}, spawnOpts)
		if err == nil || !contains(err.Error(), "darwin") {
			t.Fatalf("err = %v", err)
		}
	})
}

func hasSuffix(s, suf string) bool { return strings.HasSuffix(s, suf) }

func assertSetEqual(t *testing.T, got, want []string) {
	t.Helper()
	m := map[string]int{}
	for _, g := range got {
		m[g]++
	}
	for _, w := range want {
		m[w]--
	}
	for k, v := range m {
		if v != 0 {
			t.Fatalf("set mismatch: %q count %+v (got %v want %v)", k, v, got, want)
		}
	}
	if len(got) != len(want) {
		t.Fatalf("lengths differ: got %v want %v", got, want)
	}
}
